src/loaders/RawArrayLoader.ts

```diff
@@ -0,0 +1,138 @@
+import { Box3, Vector3 } from "three";
+
+import {
+  ThreadableVolumeLoader,
+  type LoadSpec,
+  type RawChannelDataCallback,
+  VolumeDims,
+  type LoadedVolumeInfo,
+} from "./IVolumeLoader.js";
+import { computePackedAtlasDims } from "./VolumeLoaderUtils";
+import { ImageInfo } from "../Volume";
+import { DATARANGE_UINT8 } from "../types.js";
+
+// this is the form in which a 4D numpy array arrives as converted
+// by jupyterlab into a js object.
+// This loader does not yet support multiple time samples.
+export type RawArrayData = {
+  // expected to be "uint8" always
+  dtype: "uint8";
+  // [c,z,y,x]
+  shape: [number, number, number, number];
+  // the bits (assumed uint8!!)
+  buffer: DataView;
+};
+
+// minimal metadata for visualization
+export type RawArrayInfo = {
+  name: string;
+  sizeX: number;
+  sizeY: number;
+  sizeZ: number;
+  sizeC: number;
+  physicalPixelSize: [number, number, number];
+  spatialUnit: string;
+  channelNames: string[];
+  userData?: Record<string, unknown>;
+};
+
+export interface RawArrayLoaderOptions {
+  data: RawArrayData;
+  metadata: RawArrayInfo;
+}
+
+const convertImageInfo = (json: RawArrayInfo): ImageInfo => ({
+  name: json.name,
+
+  // assumption: the data is already sized to fit in our viewer's preferred
+  // memory footprint (a tiled atlas texture as of this writing)
+  originalSize: new Vector3(json.sizeX, json.sizeY, json.sizeZ),
+  atlasTileDims: computePackedAtlasDims(json.sizeZ, json.sizeX, json.sizeY),
+  volumeSize: new Vector3(json.sizeX, json.sizeY, json.sizeZ),
+  subregionSize: new Vector3(json.sizeX, json.sizeY, json.sizeZ),
+  subregionOffset: new Vector3(0, 0, 0),
+  physicalPixelSize: new Vector3(json.physicalPixelSize[0], json.physicalPixelSize[1], json.physicalPixelSize[2]),
+  spatialUnit: json.spatialUnit || "μm",
+
+  numChannels: json.sizeC,
+  channelNames: json.channelNames,
+  channelColors: undefined, //json.channelColors,
+
+  times: 1,
+  timeScale: 1,
+  timeUnit: "s",
+
+  numMultiscaleLevels: 1,
+  multiscaleLevel: 0,
+
+  transform: {
+    translation: new Vector3(0, 0, 0),
+    rotation: new Vector3(0, 0, 0),
+  },
+
+  userData: json.userData,
+});
+
+class RawArrayLoader extends ThreadableVolumeLoader {
+  // one array per channel
+  data: RawArrayData;
```

Comment:
Is this comment still correct? It seems like `RawArrayData` has one array that merges all of the data across channels, not sure if I'm reading that incorrectly.